tests/SpecTests/PrimaryStepDownSpecTest.php

```diff
@@ -0,0 +1,278 @@
+     */
+    public function testInterruptedAtShutdownResetConnectionPool()
+    {
+        $runOn = [(object) ['minServerVersion' => '4.0.0']];
+        $this->checkServerRequirements($runOn);
+
+        // Set a fail point
+        $this->configureFailPoint([
+            'configureFailPoint' => 'failCommand',
+            'mode' => ['times' => 1],
+            'data' => [
+                'failCommands' => ['insert'],
+                'errorCode' => 11600,
+            ],
+        ]);
+
+        $totalConnectionsCreated = $this->getTotalConnectionsCreated();
+
+        // Execute an insert into the test collection of a {test: 1} document.
+        try {
+            $this->insertDocuments(1);
+        } catch (BulkWriteException $e) {
+            // Verify that the insert failed with an operation failure with 10107 code.
+            $this->assertSame(11600, $e->getCode());
+        }
+
+        // Verify that the connection pool has been cleared
+        $this->assertSame($totalConnectionsCreated + 1, $this->getTotalConnectionsCreated());
+    }
+
+    private function insertDocuments($count)
+    {
+        $operations = [];
+
+        for ($i = 1; $i <= $count; $i++) {
+            $operations[] = [
+                BulkWrite::INSERT_ONE => [['test' => $i]],
+            ];
+        }
+
+        return $this->collection->bulkWrite($operations, ['writeConcern' => new WriteConcern('majority')]);
+    }
+
+    private function dropAndRecreateCollection()
+    {
+        $this->client->selectCollection($this->getDatabaseName(), $this->getCollectionName())->drop();
+        $this->client->selectDatabase($this->getDatabaseName())->command(['create' => $this->getCollectionName()]);
+    }
+
+    private function getTotalConnectionsCreated()
+    {
+        $cursor = $this->client->getManager()->executeCommand(
+            $this->getDatabaseName(),
+            new Command(['serverStatus' => 1]),
+            new ReadPreference(ReadPreference::RP_PRIMARY)
+        );
+
+        $cursor->setTypeMap(['root' => 'array', 'document' => 'array']);
+        $document = current($cursor->toArray());
+
+        if (isset($document['connections'], $document['connections']['totalCreated'])) {
+            return (int) $document['connections']['totalCreated'];
+        }
+
+        throw new UnexpectedValueException('Could not determine number of total connections');
+    }
+
+    private function waitForMasterReelection()
+    {
+        try {
+            $this->insertDocuments(1);
+            return;
+        } catch (DriverException $e) {
+            for ($i = 0; $i < 40; $i++) {
+                try {
+                    $this->client->getManager()->selectServer(new ReadPreference('primary'));
+                    return;
+                } catch (ConnectionTimeoutException $e) {
+                    usleep(500000);
+                }
+            }
+        }
+
+        $this->fail('Expected primary to be re-elected within 20 seconds.');
```

Comment:
I don't believe the preceding loop is actually time-boxed to 20 seconds. The custom client uses the default URI, which may or may not have a custom `serverSelectionTimeoutMS` value set. I think it may make more sense to specifying the following URI array options:

 * heartbeatFrequencyMS = 500 (minimum allowed value, without patching libmongoc :wink:)
 * serverSelectionTimeoutMS = 20000 (if you really want to keep the 20-second timeout)
 * serverSelectionTryOnce = false (to allow `selectServer()` to loop internally instead of using `for` ourselves)

That should allow us to rewrite this method as:

```php
try {
    $this->insertDocuments(1);
    return;
} catch (DriverException $e) {
    $this->client->getManager()->selectServer(new ReadPreference('primary'));
}
```